Homework 4

DaniilVyatkin opened the pull request:
1. Added adapters between SensorEvent and CCSensorEvent and vice versa
2. Added adapter from basic EventHandlers to CC EventHandler
3. Added initialisation of SensorEventsManager in main in Application using @Bean and configuration from MyConfiguration

DaniilVyatkin committed:
Homework 1 on SmartHome

DaniilVyatkin committed:
Homework 2
Unfortunately, I was unable to correctly implement turning off all the lights on closing the hall door.

DaniilVyatkin committed:
Implemented 2 adapters: from CCSensorEvent to SensorEvent and vice versa.

DaniilVyatkin committed:
Implemented eventHandlerToCCEventHandlerAdapter and added initialization using @Beans from MyConfiguration in main() in Application.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,51 @@
+
+        // Я правильно понял, что теперь мы вместо нижележащего кода (4 строчки)
+        // используем SensorEventsManager из com.coolcompany.smarthome.events,
+        // который делает примерно все то же самое, но без загрузки дома?
```
Да.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MyConfiguration.java:
```diff
@@ -0,0 +1,43 @@
+        return new LightEventHandler();
+    }
+
+    @Bean
```
Не нужен `@Bean`. просто сделайте статический метод и вызывайте его в соответствующем бине: `LightEventHandler lightEventHandler() { return eventHandlerToCCEventHandlerAdapter(new LightEventHandler()) ...  }`.
Клиент не должен знать про адаптер/декоратор итд.

a-ivanov reviewed (changes requested):
Где код 3 дз? Должен быть в итоге в этой ветке.

DaniilVyatkin committed:
Homework 3
1. Создал Alarm
2. Создал интерфейс AlarmState и его реализации ActivatedAlarmState, DeactivatedAlarmState, AlertAlarmState
3. Создал AlarmEventHandler
4. Создал SecurityEventHandlerDecorator - декоратор, который оборачивает EventHandler'ы

DaniilVyatkin committed:
Homework 4 corrections
1. Удалил закомментированный код из Application
2. Исправил создание бина в MyConfiguration

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/alarm/AlertAlarmState.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void activate(int code) {
+        if (alarm.isCodeCorrect(code)) {
```
Она уже по сути активирована.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MyConfiguration.java:
```diff
@@ -0,0 +1,28 @@
+    @Bean
+    SensorEventsManager sensorEventsManager() {
+        SensorEventsManager sensorEventsManager = new SensorEventsManager();
+        sensorEventsManager.registerEventHandler(eventHandlerToCCEventHandlerAdapter(new DoorEventHandler()));
```
Создайте по `@Bean` на каждый `EventHandlerToCCEventHandlerAdapter`, здесь соберите их все в список `@Bean
    SensorEventsManager sensorEventsManager(List<EventHandler> handlers) {...}`. Смысл спринга в том, что мы можем менять реализации в настройке (конфигурации) без изменения клиентского кода и других частей конфигурации.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MyConfiguration.java:
```diff
@@ -0,0 +1,28 @@
+        return sensorEventsManager;
+    }
+
+    static EventHandlerToCCEventHandlerAdapter eventHandlerToCCEventHandlerAdapter(ru.sbt.mipt.oop.EventHandler eventHandler) {
```
Не нужно static методов.

a-ivanov replied:
Объявляйте бины по типу их интерфейса, тут ru.sbt.mipt.oop.EventHandler.
Клиентский код (SensorEventsManager) не должен знать, что вы используете адаптеры.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/EventHandlerToCCEventHandlerAdapter.java:
```diff
@@ -0,0 +1,20 @@
+import com.coolcompany.smarthome.events.CCSensorEvent;
+import com.coolcompany.smarthome.events.EventHandler;
+
+public class EventHandlerToCCEventHandlerAdapter implements EventHandler {
```
А где тут адаптер? Выглядит как декоратор, правда бессмысленный.
Вы же адаптируете `com.coolcompany.smarthome.events.EventHandler` к вашему виду обработчиков, см. https://refactoring.guru/ru/design-patterns/adapter

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/CCSensorEventToSensorEventAdapter.java:
```diff
@@ -0,0 +1,39 @@
+        String ccSensorEventEventType = ccSensorEvent.getEventType();
+        SensorEventType sensorEventType;
+        // конвертируем тип события ccSensorEventEventType в тип события SensorEventType
+        if (ccSensorEventEventType == "LightIsOn") {
```
Это все нарушение OCP, сделайте конфигурируемую мапу, мапу можно объявить в бине.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/CCSensorEventToSensorEventAdapter.java:
```diff
@@ -0,0 +1,39 @@
+
+    // пришлось вынести конвертацию типа события из конструктора в отдельную функцию, так как
+    // в конструкторе super() должно быть в первой строчке
+    private static SensorEventType CCSensorEventEventTypeToEventTypeConverter(CCSensorEvent ccSensorEvent) {
```
Не нужно здесь использовать static методов, за вас создаст инстанс спринг из бина.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/alarm/Alarm.java:
```diff
@@ -0,0 +1,36 @@
+        alarmState.alert();
+    }
+
+    public boolean isCodeCorrect(int code) {
```
package private для такого метода

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SecurityEventHandlerDecorator.java:
```diff
@@ -0,0 +1,34 @@
+            wrappedHandler.handle(event, smartHome);
+        }
+        else {
+            if (alarm.getAlarmState() instanceof AlertAlarmState) {
```
Замените getAlarmState на метод типа isAlarmed, isActivated и внутри этих методов проверяйте текущее состояние. Таким образом вы не будете показывать детали реализации, в частности, паттерн шаблон. Если клиент не знает об этом, проще будет его заменить на другую реализацию при необходимости.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/alarm/ActivatedAlarmState.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void activate(int code) {
+        if (alarm.isCodeCorrect(code)) {
```
Здесь ничего делать не надо, т.к. уже в активированном состоянии.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/SecurityEventHandlerDecorator.java:
```diff
@@ -0,0 +1,34 @@
+        }
+        else {
+            if (alarm.getAlarmState() instanceof AlertAlarmState) {
+                System.out.println("Alarm is in alert state and caught some sensor activity!");
```
Здесь нарушение SRP, вынесите функциональность по оповещению в отдельный интерфейс Notifier с реализацией DummySMSNotifier.
Тут добавьте поле Notifier, передавайте в конфигурации спринга бин с реализацией DummySMSNotifier.

DaniilVyatkin committed:
Homework 3 corrections
1. Сделал package private для isCodeCorrect в Alarm
2. Заменил getAlarmState на isActivated, isDeactivated, isAlert в Alarm
3. Убрал действия из ActivatedAlarmState при попытке повторной активации
4. Вынес отправку sms из SecurityEventHandlerDecorator в интерфейс Notifier и его реализацию DummySMSNotifier

DaniilVyatkin committed:
Homework 4 corrections
1. Убрал действия при попытке активировать сигнализацию в состоянии тревоги в AlertAlarmState
2. Создайте по @Bean на каждый EventHandler, реализовал их сбор в список в sensorEventManager
3. Исправил использование getEventHandlerAdaptedToCC
4. Исправил EventHandlerToCCEventHandlerAdapter
5. Добавил в конфигурацию мапу для переводов типов событий из CC в SensorEvent

DaniilVyatkin committed:
Homework 5 inserted into Homework 4
Перенес все из Homework-5 в Homework-4, чтобы весь актуальный код был в одной ветке

DaniilVyatkin committed:
Homework 5 Corrections
1. Сигнализация из 3 дз
2. Перенес настройку кнопок пульта в конфигурацию
3. Переделал конфигурацию SensorEventsManager по замечаниям к 4 дз
4. Исправил конфигурацию RemoteControlRegistry с добавлением пульта
5. Добавил id в конструктор CloseDoorAction
6. Исправил передачу door.getId() в CloseHallDoorCommandRC
7. Убрал getRooms из SmartHome и других классов, использовавших этот метод
